packages/app/src/systems/Core/components/CoinBalance.tsx

```diff
@@ -0,0 +1,55 @@
+import { forwardRef, useMemo } from "react";
+
+import { useBalances } from "../hooks";
+import { parseToFormattedNumber } from "../utils";
+
+import { Button, Tooltip } from "~/systems/UI";
+import type { Coin } from "~/types";
+
+export type CoinBalanceProps = {
+  coin?: Coin | null;
+  showMaxButton?: boolean;
+  onSetMaxBalance?: () => void;
+  gasFee?: bigint | null;
+};
+
+export const CoinBalance = forwardRef<HTMLDivElement, CoinBalanceProps>(
+  ({ coin, showMaxButton = true, onSetMaxBalance, gasFee }) => {
+    const { data: balances } = useBalances({ enabled: true });
+    const coinBalance = balances?.find(
+      (item) => item.assetId === coin?.assetId
+    );
+
+    const balance = useMemo(
```

Comment:
You can add `coinBalance` inside the `useMemo` as well, to avoid `find` on each re-render:

```js
const balance = useMemo(() => {
  const coinBalance = balances?.find((item) => item.assetId === coin?.assetId);
  return parseToFormattedNumber(coinBalance?.amount || BigInt(0)),
}, [coin?.assetId])
```